### PR TITLE
Lazy string evaluation for exceptions to improve performance (closes #1375)

### LINKED ISSUE #1375: NEMO raising performance problems
I've succeeded in making the parsing overhead look small because the transformations that raise language-level PSyIR to NEMO PSyIR can take a very long time if a routine contains a lot of code (e.g. for lbcnfd.f90 and lib_mpp.f90 in NEMO).

I've alleviated some of this in #1147 by merging two `walk()` calls into one but the situation is still not good. It's possible that adding an option to `walk()` so that it returns nodes depth first would improve things as then we could transform the deepest loops first.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by sergisiso** [comment]
@rupertford looks good 👍

**Comment by rupertford** [comment]
Ready for first review from @sergisiso or @arporter subject to GHA CI happiness

File: src/psyclone/tests/psyir/transformations/transformation_error_test.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+'''pytest tests for the transformation_errors module.'''
+
```
Comment (sergisiso) [inline src/psyclone/tests/psyir/transformations/transformation_error_test.py:38]: pylint says "import missing `from __future__ import absolute_import`"

File: src/psyclone/tests/errors_test.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+'''pytest tests for the errors module.'''
+
```
Comment (sergisiso) [inline src/psyclone/tests/errors_test.py:38]: pylint says "import missing `from __future__ import absolute_import`"

File: src/psyclone/errors.py [hunk 3]
```diff
@@ -39,6 +39,37 @@
...
+    def __init__(self, func):
+        if not hasattr(func, '__call__'):
+            raise TypeError(
+                "The func argument for the LazyString class should be a "
+                "function, but found '{0}'.".format(type(func).__name__))
+        self._func = func
+
+    def __str__(self):
+        '''
+        :raises TypeError: if the the function stored in self._func does \
```
Comment (sergisiso) [inline src/psyclone/errors.py:62]: s/the the/the/

File: src/psyclone/errors.py [hunk 4]
```diff
@@ -39,6 +39,37 @@
 ''' This module provides various error classes used in PSyclone'''
 
 
+class LazyString:
```
Comment (sergisiso) [inline src/psyclone/errors.py:42]: Add class to the `__all__` list at the bottom.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@rupertford Looks good, some small fixes and this can be merged.

File: src/psyclone/tests/psyir/transformations/transformation_error_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/transformation_error_test.py:38]: Added. Apologies, I don't get that for my version of pylint (2.10.2)

File: src/psyclone/tests/errors_test.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/tests/errors_test.py:38]: As above.

File: src/psyclone/errors.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/errors.py:62]: Oops.

File: src/psyclone/errors.py (hunk 4 above)
Comment (rupertford) [reply src/psyclone/errors.py:42]: Oops.

**Comment by rupertford** [comment]
Read for next review from @sergisiso

**Review by sergisiso — APPROVED** [review]
@rupertford All comments addressed.
- No remaining TODOs.
- Compilations pass with tests and diff coverage 100%
- pycodestyle and pylint don't show new relevant issues.
- New functionality is properly documented and documentation builds properly.

This is ready to merge.

(I was playing with another implementation where the Error constructors are not overrider so the LazyString would be accepted by default without explicit support in each error class but this requires #1389, so I left this idea for future discussion)

**Pull request merged**